### PR TITLE
Capacitors and some item refactoring.

### PR DESCRIPTION
**Pull request by Rover656**
# Description

Refactors some items and brings in #10 

# Checklist:

- [x] My code follows the style guidelines of this project (.editorconfig, most IDEs will use this for you)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: endercore/src/main/java/com/enderio/core/common/capability/MultiCapabilityProvider.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Automatically combines multiple capabilities together; handling serialization too.
+ */
+public class MultiCapabilityProvider implements ICapabilitySerializable<CompoundTag> {
+    // TODO: Allow non-serialized too...
```
Comment (Rover656) [inline endercore/src/main/java/com/enderio/core/common/capability/MultiCapabilityProvider.java]: We should have a discussion about this; my knowledge regarding capabilities is limited, so I think this todo should be dealt with before merge

File: endercore/src/main/java/com/enderio/core/common/capability/MultiCapabilityProvider.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    @Override
+    public void deserializeNBT(CompoundTag nbt) {
+        for (var entry : capabilities.entrySet()) {
+            entry
+                .getValue()
+                .ifPresent(capability -> {
+                    String key = entry
+                        .getKey()
+                        .getName(); // TODO: More robust keys? Moving a package or renaming a class will break this completely.
```
Comment (Rover656) [inline endercore/src/main/java/com/enderio/core/common/capability/MultiCapabilityProvider.java]: Could do with suggestions; not really the kind of thing we'd want going into production.

File: endercore/src/main/java/com/enderio/core/mixin/common/MultiCapProviderMixin.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.world.item.ItemStack;
+import net.minecraftforge.common.capabilities.ICapabilityProvider;
+import net.minecraftforge.common.extensions.IForgeItem;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+
+import javax.annotation.Nullable;
+
+@Mixin(IMultiCapability.class)
```
Comment (Rover656) [inline endercore/src/main/java/com/enderio/core/mixin/common/MultiCapProviderMixin.java]: Add comment describing the mixin's purpose and how it works. Not strictly necessary but I believe we should maybe maintain good documentation across the board, but to start we should comment things that are slightly less obvious; for example how a mixin works, why its there and what its used for?

File: enderio-base/src/main/java/com/enderio/base/common/capability/EIOCapabilities.java [hunk 4]
```diff
@@ -11,14 +14,24 @@
 
 @Mod.EventBusSubscriber(modid = EnderIO.MODID, bus = Mod.EventBusSubscriber.Bus.MOD)
 public class EIOCapabilities {
-    
+    @CapabilityInject(IEntityStorage.class)
```
Comment (Rover656) [inline enderio-base/src/main/java/com/enderio/base/common/capability/EIOCapabilities.java]: Move away from CapabilityInject like the main branch

File: enderio-base/src/main/java/com/enderio/base/common/capability/entity/EntityStorage.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
+package com.enderio.base.common.capability.entity;
+
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.Tag;
+import net.minecraft.resources.ResourceLocation;
+
+import javax.annotation.Nonnull;
+import java.util.Optional;
+
+public class EntityStorage implements IEntityStorage {
```
Comment (Rover656) [inline enderio-base/src/main/java/com/enderio/base/common/capability/entity/EntityStorage.java:10]: Document it

File: enderio-base/src/main/java/com/enderio/base/common/capability/entity/IEntityStorage.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.Tag;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraftforge.common.util.INBTSerializable;
+
+import javax.annotation.Nonnull;
+import java.util.Optional;
+
+public interface IEntityStorage extends INBTSerializable<Tag> {
```
Comment (Rover656) [inline enderio-base/src/main/java/com/enderio/base/common/capability/entity/IEntityStorage.java]: Document

File: enderio-base/src/main/java/com/enderio/base/common/capability/toggled/IToggled.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
+package com.enderio.base.common.capability.toggled;
+
+import net.minecraft.nbt.Tag;
+import net.minecraftforge.common.util.INBTSerializable;
+
+import javax.annotation.Nonnull;
+
+public interface IToggled extends INBTSerializable<Tag> {
```
Comment (Rover656) [inline enderio-base/src/main/java/com/enderio/base/common/capability/toggled/IToggled.java]: Document me

File: enderio-base/src/main/java/com/enderio/base/common/capability/toggled/Toggled.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
+package com.enderio.base.common.capability.toggled;
+
+import net.minecraft.nbt.ByteTag;
+import net.minecraft.nbt.Tag;
+
+public class Toggled implements IToggled {
```
Comment (Rover656) [inline enderio-base/src/main/java/com/enderio/base/common/capability/toggled/Toggled.java:6]: docs

File: enderio-base/src/main/java/com/enderio/base/common/item/EIOItems.java [hunk 9]
```diff
@@ -83,6 +93,35 @@
...
+    public static final ItemEntry<Item> BASIC_CAPACITOR = REGISTRATE.item("basic_capacitor", Item::new).properties(p -> p.stacksTo(1)).group(() -> EIOCreativeTabs.MAIN).register();
+    public static final ItemEntry<Item> DOUBLE_LAYER_CAPACITOR = REGISTRATE.item("double_layer_capacitor", Item::new).properties(p -> p.stacksTo(1)).group(() -> EIOCreativeTabs.MAIN).register();
+    public static final ItemEntry<Item> OCTADIC_CAPACITOR = REGISTRATE.item("octadic_capacitor", Item::new).properties(p -> p.stacksTo(1)).group(() -> EIOCreativeTabs.MAIN).register();
+    public static final ItemEntry<LootCapacitorItem> LOOT_CAPACITOR = REGISTRATE.item("loot_capacitor", LootCapacitorItem::new).properties(p -> p.stacksTo(1)).register();
+    
+    // endregion
+    
+    // region capacitor text
+    
+    public static final TranslatableComponent ALL_ENERGY_CONSUMPSTION = capacitorDescriptionBuilder("type", ICapacitorData.ALL_ENERGY_CONSUMPSTION, "Leaky");
```
Comment (Rover656) [inline enderio-base/src/main/java/com/enderio/base/common/item/EIOItems.java:126]: Add a todo here; "TODO: Move language related things to a dedicated location, like EIOLang"

File: enderio-base/src/main/java/com/enderio/base/common/item/capacitors/LootCapacitorItem.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+import net.minecraftforge.common.util.LazyOptional;
+
+public class LootCapacitorItem extends Item implements IMultiCapability{
+    private CapacitorData data = new CapacitorData();
+    
+    public LootCapacitorItem(Properties pProperties) {
+        super(pProperties);
+    }
+    
+    public LootCapacitorItem(Properties pProperties, float base, Map<String, Float> specialization) {
```
Comment (Rover656) [inline enderio-base/src/main/java/com/enderio/base/common/item/capacitors/LootCapacitorItem.java]: Is this required anymore, all loot capacitors generate their specialisations when they are generated

File: enderio-base/src/main/java/com/enderio/base/common/item/tool/SoulVialItem.java [hunk 11]
```diff
@@ -64,6 +72,8 @@ public void appendHoverText(@Nonnull ItemStack pStack, @Nullable Level pLevel, @
     @Override
     public InteractionResult interactLivingEntity(@Nonnull ItemStack pStack, @Nonnull Player pPlayer, @Nonnull LivingEntity pInteractionTarget,
         @Nonnull InteractionHand pUsedHand) {
+        // TODO: Can't pickup donkey?
```
Comment (Rover656) [inline enderio-base/src/main/java/com/enderio/base/common/item/tool/SoulVialItem.java]: Fix this before merge

File: enderio-base/src/main/java/com/enderio/base/common/item/tool/electromagnet/ElectromagnetItem.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+import net.minecraft.world.item.CreativeModeTab;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraftforge.common.util.LazyOptional;
+import net.minecraftforge.energy.CapabilityEnergy;
+import net.minecraftforge.energy.EnergyStorage;
+
+import javax.annotation.Nullable;
+
+// TODO: Behaviours
```
Comment (Rover656) [inline enderio-base/src/main/java/com/enderio/base/common/item/tool/electromagnet/ElectromagnetItem.java:22]: Should this be finished before merge?

File: enderio-base/src/main/java/com/enderio/base/common/item/util/IEnergyBar.java [hunk 13]
```diff
@@ -0,0 +1,6 @@
+package com.enderio.base.common.item.util;
+
+/**
+ * Add to any item that must display an energy bar on the item stack.
```
Comment (Rover656) [inline enderio-base/src/main/java/com/enderio/base/common/item/util/IEnergyBar.java:4]: Explain how the interface actually works

File: enderio-base/src/main/java/com/enderio/base/common/recipe/CapacitorDataRecipe.java [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+    public ICapacitorData getCapacitorData() {
+        return capacitorData;
+    }
+
+    public boolean matches(ItemStack stack) {
+        return capacitor.test(stack);
+    }
+
+    //should prevent Unknown RecipeCategory log spamming
```
Comment (Rover656) [inline enderio-base/src/main/java/com/enderio/base/common/recipe/CapacitorDataRecipe.java]: Validate whether this is true, if so then remove the uncertainty from the comment

File: enderio-base/src/main/java/com/enderio/base/common/recipe/DataGenSerializer.java [hunk 15]
```diff
@@ -0,0 +1,10 @@
+package com.enderio.base.common.recipe;
+
+import com.google.gson.JsonObject;
+import net.minecraft.world.Container;
+import net.minecraft.world.item.crafting.Recipe;
+import net.minecraft.world.item.crafting.RecipeSerializer;
+
+public abstract class DataGenSerializer<T extends Recipe<C>, C extends Container> extends net.minecraftforge.registries.ForgeRegistryEntry<RecipeSerializer<?>> implements RecipeSerializer<T> {
```
Comment (Rover656) [inline enderio-base/src/main/java/com/enderio/base/common/recipe/DataGenSerializer.java]: Name isn't clear enough

File: enderio-base/src/main/java/com/enderio/base/data/model/item/ItemModelUtils.java [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+            .texture("all", prov.itemTexture(ctx));
+    }
+
+    public static void mimicItem(DataGenContext<Item, ? extends Item> ctx, ItemEntry<? extends Item> item, RegistrateItemModelProvider prov) {
+        prov.generated(ctx, prov.itemTexture(item));
+    }
+
+    public static void gearItem(DataGenContext<Item, ? extends Item> ctx, RegistrateItemModelProvider prov) {
+        // json so the BEWLR is used + perspectives
+        prov
```
Comment (Rover656) [inline enderio-base/src/main/java/com/enderio/base/data/model/item/ItemModelUtils.java]: Formatting needs looked at

File: enderio-base/src/main/java/com/enderio/base/mixin/client/EnergyBarMixin.java [hunk 17]
```diff
@@ -0,0 +1,28 @@
+package com.enderio.base.mixin.client;
+
+import com.enderio.base.common.item.util.IEnergyBar;
+import net.minecraft.world.item.ItemStack;
+import net.minecraftforge.common.extensions.IForgeItem;
+import net.minecraftforge.energy.CapabilityEnergy;
+import org.spongepowered.asm.mixin.Mixin;
+
+@Mixin(IEnergyBar.class)
```
Comment (Rover656) [inline enderio-base/src/main/java/com/enderio/base/mixin/client/EnergyBarMixin.java:16]: Again, document the ins-and-outs of all mixins

**Review by Rover656 — COMMENTED** [review]
Adding a bunch of comments to spark discussions and to let myself know what needs done :)

File: enderio-base/src/main/java/com/enderio/base/common/item/tool/SoulVialItem.java (hunk 11 above)
Comment (Rover656) [reply enderio-base/src/main/java/com/enderio/base/common/item/tool/SoulVialItem.java]: This is going to need a mixin or something in forge to fix due to how donkeys react to interaction. Will likely just add a mixin but I hate that that's required.

**Review by justliliandev — APPROVED** [review]
The MultiCapabilityItem and Provider look fine now

File: enderio-base/src/main/java/com/enderio/base/common/capability/entity/EntityStorage.java (hunk 5 above)
Comment (Rover656) [reply enderio-base/src/main/java/com/enderio/base/common/capability/entity/EntityStorage.java:10]: ignoring, only documenting the main capability interface

File: enderio-base/src/main/java/com/enderio/base/common/capability/toggled/Toggled.java (hunk 8 above)
Comment (Rover656) [reply enderio-base/src/main/java/com/enderio/base/common/capability/toggled/Toggled.java:6]: ignoring, only documenting the main capability interface

File: enderio-base/src/main/java/com/enderio/base/common/recipe/CapacitorDataGenSerializer.java [hunk 18]
```diff
@@ -0,0 +1,12 @@
+package com.enderio.base.common.recipe;
+
+import com.google.gson.JsonObject;
+import net.minecraft.world.Container;
+import net.minecraft.world.item.crafting.Recipe;
+import net.minecraft.world.item.crafting.RecipeSerializer;
+import net.minecraftforge.registries.ForgeRegistryEntry;
+
+public abstract class CapacitorDataGenSerializer<T extends Recipe<C>, C extends Container> extends ForgeRegistryEntry<RecipeSerializer<?>>
```
Comment (justliliandev) [inline enderio-base/src/main/java/com/enderio/base/common/recipe/CapacitorDataGenSerializer.java]: can you rename it back? The DataGenSerilizer was meant for all recipes we might wanna add.
for that you would need to fix the type of the first parameter of toJson from CapacitorDataRecipe to the generic type T

**Pull request merged**